### PR TITLE
block.json: Allow passing filename as `variations` field

### PR DESCRIPTION
**Pull request by ockham**
Modeled after https://core.trac.wordpress.org/changeset/54132/, which introduced the `render` field for `block.json`, as a way to point to a PHP file instead of providing a `render_callback`.

The main difference to the `variations` and `variation_callback` fields is that the `variations` field already existed prior to this PR, and it can be used to provide the static list of variations (i.e., an array). This PR makes it so that the field can be alternatively set to a string, which will be interpreted as the filename of the PHP file generating the variations.

Trac ticket: https://core.trac.wordpress.org/ticket/61280

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: tests/phpunit/tests/blocks/register.php [hunk 1]
```diff
@@ -957,6 +957,37 @@ public function data_register_block_registers_with_args_override_returns_false_w
...
+		};
+
+		add_filter( 'block_type_metadata', $filter_metadata_registration, 10, 2 );
+		$result = register_block_type_from_metadata(
+			DIR_TESTDATA . '/blocks/notice'
+		);
+		remove_filter( 'block_type_metadata', $filter_metadata_registration );
+
+		$this->assertInstanceOf( 'WP_Block_Type', $result, 'The block was not registered' );
+		$this->assertSame( 'variations.php', $result->variations, 'The block variations are incorrect' );
```
Comment (gziolo) [inline tests/phpunit/tests/blocks/register.php]: That’s rather unwanted behavior as variations is expected to be an array on the block type instance. It should be never set to the file path. I’m also not entirely sure how it works exactly as I would assume it internally calls:

https://github.com/WordPress/wordpress-develop/blob/fa7902c8c85a2d41b801df6717e5067af8396cd3/src/wp-includes/class-wp-block-type.php#L364L367

So technically speaking, calling manually `variations_callback` later in the test wouldn’t be necessary.

**Review by gziolo — COMMENTED** [review]
That looks like nice enhancement. We will need some changes also on the editor side:
- extend block.json schema to cover the string value
- add simple logic that ignores the file path when consuming the metadata from JS as it can’t process PHP file - for WordPress context it’s going to be handled on the server anyway

File: tests/phpunit/tests/blocks/register.php (hunk 1 above)
Comment (ockham) [reply tests/phpunit/tests/blocks/register.php]: That's a good point. As we briefly discussed in person, I'll change the implementation to unset `variations` before creating the `WP_Block` instance, and the test to check that the output of `$result->variations` (and/or `->get_variations()` is indeed the array created by the PHP file).

Comment (ockham) [reply tests/phpunit/tests/blocks/register.php]: https://github.com/WordPress/wordpress-develop/pull/6668/commits/72a0384dd34b99e8a8b55b0b752e7a2949916ac9

**Comment by ockham** [comment]
> That looks like nice enhancement. We will need some changes also on the editor side:
> 
> * extend block.json schema to cover the string value

That's handled in https://github.com/WordPress/gutenberg/pull/62092. I'm going to extend that PR to also implement the logic from this PR in the 6.7 compat layer, and to include changes to block variation docs.

> * add simple logic that ignores the file path when consuming the metadata from JS as it can’t process PHP file - for WordPress context it’s going to be handled on the server anyway

👍

File: tests/phpunit/tests/blocks/register.php (hunk 1 above)
Comment (gziolo) [reply tests/phpunit/tests/blocks/register.php]: The new implementation looks and tests great 👍🏻

File: tests/phpunit/tests/blocks/register.php [hunk 2]
```diff
@@ -957,6 +957,37 @@ public function data_register_block_registers_with_args_override_returns_false_w
...
+	/**
+	 * Tests registering a block with variations from a PHP file.
+	 *
+	 * @ticket 61280
+	 *
+	 * @covers ::register_block_type_from_metadata
+	 */
+	public function test_register_block_type_from_metadata_with_variations_php_file() {
+		$filter_metadata_registration = static function ( $metadata ) {
+			$metadata['variations'] = 'variations.php';
```
Comment (gziolo) [inline tests/phpunit/tests/blocks/register.php]: Should we promote using the `file:` prefix for consistency?

```suggestion
			$metadata['variations'] = 'file:variations.php';
```

It initially came from block assets – CSS and JS file that devs can define in two forms:
- file path: `file:index.js`
- script/style handle: `my-registered-script`

In this case we don't need the distinction for now, but maybe we should make it more strict.

**Review by gziolo — APPROVED** [review]
I had one minor comment regarding enforcing `file:` prefix for paths. However, conceptually everything looks great. I like how this proposal abstracts the usage of lazy evaluated definition of block variations. 

It even made me wonder, whether we should do a similar trick with `variations_callback` for the block variations defined as an array in `block.json`.

File: tests/phpunit/tests/blocks/register.php (hunk 2 above)
Comment (ockham) [reply tests/phpunit/tests/blocks/register.php]: Yeah, makes sense! Turns out the `file:` prefix is also used by plenty of blocks in their `render` fields: https://github.com/search?q=repo%3AWordPress%2Fgutenberg%20%22file%3A.%2Frender.php%22&type=code

Comment (ockham) [reply tests/phpunit/tests/blocks/register.php]: 9a30fb28f6

Comment (gziolo) [reply tests/phpunit/tests/blocks/register.php]: Excellent 👍

**Comment by ockham** [comment]
> I had one minor comment regarding enforcing `file:` prefix for paths. However, conceptually everything looks great. I like how this proposal abstracts the usage of lazy evaluated definition of block variations.

Thank you! I [changed](https://github.com/WordPress/wordpress-develop/commit/9a30fb28f6fe6ac1b2a04b7b05ce8764f71047cd) the unit test [according to your suggestion](https://github.com/WordPress/wordpress-develop/pull/6668#discussion_r1644246809) 😊 (The code I was using for the `variations` field was copied from the `render` field, so it already had a call to `remove_block_asset_path_prefix`, which processed the `file:` prefix.)

Since we're still in Beta, I'm planning to wait for RC1 (after which a branch for 6.6 should be created) before I commit this change to `trunk`.

> It even made me wonder, whether we should do a similar trick with `variations_callback` for the block variations defined as an array in `block.json`.

You mean so that all variations aren't loaded and built right away if they're specified in `block.json`? 🤔 Yeah, we could do that. FWIW, I was planning to allow passing a JSON file as `variations` value, which would allow block authors to simply move a static variations definition to a separate file. This will likely require code that's quite similar.

**Comment by gziolo** [comment]
> Since we're still in Beta, I'm planning to wait for RC1 (after which a branch for 6.6 should be created) before I commit this change to trunk.

Yes, it sounds like a great plan.

> You mean so that all variations aren't loaded and built right away if they're specified in block.json? 🤔 Yeah, we could do that. FWIW, I was planning to allow passing a JSON file as variations value, which would allow block authors to simply move a static variations definition to a separate file. This will likely require code that's quite similar.

Yes, that would be an enhancement specific only to `block.json` integration. Referencing a JSON file also sounds like a good addition. The only caveat is that it will need some coordination with WP-CLI team to ensure that translations are still handled for variations when moved to another JSON file.

**Comment by ockham** [comment]
Thank you @gziolo!

I'll have to re-review this PR now that I've landed the [GB counterpart](https://github.com/WordPress/gutenberg/pull/62092), to see if any changes need carrying over.

**Comment by ockham** [comment]
> I'll have to re-review this PR now that I've landed the [GB counterpart](https://github.com/WordPress/gutenberg/pull/62092), to see if any changes need carrying over.

Alright, looks like it doesn't need any changes. I'll land it 😊

**Comment by ockham** [comment]
Committed to Core in https://core.trac.wordpress.org/changeset/58801.

**Comment by noisysocks** [comment]
Thanks for committing this early in the 6.7 cycle ❤️

**Pull request closed without merging**